### PR TITLE
Support Scala.js 1.0.0

### PR DESCRIPTION
**Pull request by milessabin**
We now cross build for Scala.js 0.6.32 and 1.0.0.

In consequence, drop support for Scala 2.10.x. Special support for 2.13.0-M4 was dropped at the same time.

The update to SJS 1.0.0 revealed a bug and implementation issues in the macro support for `FromProductArgs` which are corrected in the first commit of this PR.

Supersedes #959 ... many thanks to @cquiroz for his assistance with this.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -256,32 +256,18 @@ lazy val scalaMacroDependencies: Seq[Setting[_]] = Seq(
   libraryDependencies ++= Seq(
     scalaOrganization.value % "scala-reflect" % scalaVersion.value % "provided",
     scalaOrganization.value % "scala-compiler" % scalaVersion.value % "provided"
-  ),
-  libraryDependencies ++= {
-    CrossVersion.partialVersion(scalaVersion.value) match {
-      // macro-compat and Macro paradise needed for Scala 2.10
-      case Some((2, 10)) => Seq(
```
Comment (cquiroz) [inline build.sbt]: Removing 2.10 does clean the build up 😄

File: core/src/main/scala/shapeless/hlists.scala [hunk 2]
```diff
@@ -373,12 +386,4 @@ class ProductMacros(val c: whitebox.Context) extends SingletonTypeUtils with Nat
         c.abort(c.enclosingPosition, s"Expression $elem does not evaluate to a non-negative Int literal")
     }._2
   }
-
-  def mkParamsImpl[L <: HList](method: MethodSymbol, product: Expr[L]): List[List[Tree]] = {
-    val slices = method.paramLists.filterNot(_.forall(x => x.isImplicit))
-      .foldLeft((List[List[Int]](), 0))((acc, e) =>
-        (acc._1 :+ (acc._2 to (acc._2 + (e.size - 1))).toList, acc._2 + e.size))._1
-
-    slices.map(_.map(i => q"${product}.apply(${i})"))
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/hlists.scala]: Yikes, this was in desperate need of rewriting 👍

File: README.md [hunk 3]
```diff
@@ -132,17 +132,6 @@ libraryDependencies ++= Seq(
 )
 ```
 
-If you are using Scala 2.10.x, you should also add the macro paradise plugin to your build,
-
-```scala
-scalaVersion := "2.10.7"
```
Comment (joroKr21) [inline README.md]: So much better without Scala 2.10

**Comment by SethTisue** [comment]
@milessabin I think a lot of people are blocked waiting on actual publishing of this. Are you blocked on some technical issue, or just haven't been able to find the time quite yet?

**Comment by milessabin** [comment]
I've just back-published 2.3.3 for 2.{11,12,13}.

**Pull request merged**